### PR TITLE
Move dynamodb session setup into per-request lifecycle

### PR DESCRIPTION
**Pull request by nealedj**
### What is the context of this PR?
This removes the setup of the dynamodb session from the app setup as this is recommended to me per thread.

### How to review 
- Check that question playback still functions

### CODE REVIEW

File: app/globals.py [hunk 1]
```diff
@@ -62,3 +65,23 @@ def get_answer_store(user):
...
+
+    db = getattr(g, '_dynamodb', None)
+    if db is None:
+        # Number of additional connection attempts
+        config = Config(retries={'max_attempts': 1})
+
+        dynamodb = boto3.resource(
+            'dynamodb', endpoint_url=current_app.config['EQ_DYNAMODB_ENDPOINT'], config=config)
+
+        db = g._dynamodb = dynamodb
```
Comment (jonnyshaw89) [inline app/globals.py]: Would `return g._dynamodb = dynamodb` work?

File: tests/integration/views/test_view_submission.py [hunk 2]
```diff
@@ -47,18 +44,20 @@ def test_view_submission(self):
...
 
         # Submit answers
-        self.table.put_item = Mock(return_value={'ResponseMetadata': {'HTTPStatusCode': 200}})
-        self.post(action=None)
+        with mock.patch('app.data_model.submitted_responses.put_item',
+                        return_value={'ResponseMetadata': {'HTTPStatusCode': 200}}) as m_put_item:
+            self.post(action=None)
 
-        item = self.table.put_item.call_args[1]
+            item = m_put_item.call_args[0][0]
```
Comment (jonnyshaw89) [inline tests/integration/views/test_view_submission.py]: anything wrong with just `m_put_item` i see the intention behind it but it brakes consistency

**Review by jonnyshaw89 — CHANGES_REQUESTED** [review]

File: app/globals.py (hunk 1 above)
Comment (nealedj) [reply app/globals.py]: Yes, but that line only gets reached if the dynamo resource object wasn't already on the context object. So we'd have two returns:

```
db = getattr(g, '_dynamodb', None)
if db is None:
   dynamodb = stuff()
   return g._dynamodb = dynamodb
return db
```

File: tests/integration/views/test_view_submission.py (hunk 2 above)
Comment (nealedj) [reply tests/integration/views/test_view_submission.py]: If we use `m_put_item` outside of the patch context manager then we'll need to indent the following patch inside of it.

I.e.
```
with mock.patch('put_item') as m_put_item:
    ...stuff
    with mock.patch('get_item', return_value=item):
        ...stuff
```

File: app/globals.py (hunk 1 above)
Comment (jonnyshaw89) [reply app/globals.py]: Ah yeh I suppose its tidier. Just seems odd to set a variable that we don't need.

File: tests/integration/views/test_view_submission.py (hunk 2 above)
Comment (jonnyshaw89) [reply tests/integration/views/test_view_submission.py]: i'm not sure I follow.

**Review by jonnyshaw89 — APPROVED** [review]

File: app/data_model/submitted_responses.py [hunk 3]
```diff
@@ -1,37 +1,44 @@
...
-                item['valid_until'] = datetime.utcfromtimestamp(item['valid_until'])
-            return item
+def _get_table():
+    dynamodb = get_dynamodb()
+    table_name = current_app.config['EQ_SUBMITTED_RESPONSES_TABLE_NAME']
 
-        except (BotoCoreError, BotoConnectionError, ConnectionError):
-            logger.warn('Could not get data from DynamoDB')
-            return None
+    return dynamodb.Table(table_name)
```
Comment (dwyeradam) [inline app/data_model/submitted_responses.py]: Wouldn't this fail if get_dynamodb() returns None? I realise that currently everything checks that dynamodb is enabled before calling put_item or get_item but should we have some checks within submitted_responses.py to ensure this? 
See: https://github.com/ONSdigital/eq-survey-runner/pull/1486/files#diff-a160493e40efb08443761eff2a589ea2R72

**Review by dwyeradam — CHANGES_REQUESTED** [review]

File: app/data_model/submitted_responses.py (hunk 3 above)
Comment (nealedj) [reply app/data_model/submitted_responses.py]: Good catch!

**Review by dwyeradam — CHANGES_REQUESTED** [review]
Happy with the code. These changes have dropped the code coverage though so the tests may need to be reworked to hit more.

File: tests/app/data_model/test_submitted_responses.py [hunk 4]
```diff
@@ -20,36 +17,58 @@ def setUp(self):
...
+    @mock.patch('app.data_model.submitted_responses._get_table', return_value=None)
+    def test_no_table(self, _get_table):
+        self.assertFalse(submitted_responses.put_item(self.item))
+        self.assertFalse(submitted_responses.get_item('test_tx_id'))
+
+
+class SubmittedResponsesGetTableTest(unittest.TestCase):
+    @mock.patch('app.data_model.submitted_responses.get_dynamodb')
+    @mock.patch('app.data_model.submitted_responses.current_app', **{'config.new': {'EQ_SUBMITTED_RESPONSES_TABLE_NAME': 'foo'}})
+    def test_get_table(self, get_dynamodb, current_app): # pylint: disable=unused-argument
```
Comment (jonnyshaw89) [inline tests/app/data_model/test_submitted_responses.py:73]: what is this actually doing?

Comment (nealedj) [reply tests/app/data_model/test_submitted_responses.py:73]: Testing that an object is returned by `_get_table` if `get_dynamodb()` has returned the dynamo resource and there is a `table_name` defined. Which is basically all this method does.

Also just makes sure the function is run as part of the test suite.

**Review by dwyeradam — APPROVED** [review]

**Comment by dwyeradam** [comment]
@nealedj can be squashed and we'll merge it after.

**Comment by jonnyshaw89** [comment]
@shahi645 we don't use the "update" button as it causes messy commit history

**Comment by shahi645** [comment]
ahh okay, my bad

**Pull request merged**